### PR TITLE
Cleanup /data/tmp/ci/dev_preview of files older than 30 days and empty directories

### LINKED ISSUE #143: Be sure that we correctly cleanup "dev_preview" archives build by kiwix-build.
This issue probably trigger a change or script creation outside of kiwix-build. But as it is kiwix-build creating the archive, I create the issue here. Move it in another project if you think there is a better place.

kiwix-build's PR now publish "dev preview" archives in https://tmp.kiwix.org/ci/dev_preview. This is nice to test binaries and such but it will take too much space with time.

We have to cleanup the directory time to time.
I suspect a cron job which remove directories older than one month should be enough.

### PR DESCRIPTION
**Pull request by benoit74**
Fix #143 

Changes:
- create the requested cronjob
- reuse same image as tmp-kiwix pod so that it is most probably already present (we just need `find` and `wc`)
- in a single command, the job runs:
  - a dry run of files to delete to count items before deleting them
  - a deletion of files to delete
  - a dry run of directories to delete to count items before deleting them
  - a deletion of directories to delete

### CODE REVIEW

**Review by rgaudin — CHANGES_REQUESTED** [review]
Thank you.

I have a few general comments:

- I'd prefer we use our [*maintenance*](https://github.com/kiwix/k8s/blob/ec22cbecc50316d69020ae411690e4575b2ff885/zim/library-mgmt/maintenance.job.yaml#L12) image instead of an _nginx_ one (?!). It allows automatically downloading/installing scripts from this repo which makes maintenance a lot easier.
- I don't see the point of the “dry-run” commands If the goal is to display the files, appending `-print` to the find command does it. “dry-run” is misleading in this case as the delete command follows right after so it prevents nothing.
- Such destructive scripts should be using `set -e` so that any failing command blocks the rest of the script.
- I'd actually prefer we write this script in Python. We can do it easily with the stdlib and we have access to python so expecting maintainers to master shell scripting in addition to python seems superfluous.

**Comment by benoit74** [comment]
Great, thank you!

> I'd prefer we use our [_maintenance_](https://github.com/kiwix/k8s/blob/ec22cbecc50316d69020ae411690e4575b2ff885/zim/library-mgmt/maintenance.job.yaml#L12) image instead of an _nginx_ one (?!). It allows automatically downloading/installing scripts from this repo which makes maintenance a lot easier.

That's cool, how to use it? Do I get it right that it means you expect the cleanup script to live in this k8s repo? Any recommendation in term of location? (same folder than k8s cronjob?). Btw, the nginx image is the one used by tmp-kiwix, as mentioned, but never mind.

> I don't see the point of the “dry-run” commands If the goal is to display the files, appending `-print` to the find command does it. “dry-run” is misleading in this case as the delete command follows right after so it prevents nothing.

You are right.

> Such destructive scripts should be using `set -e` so that any failing command blocks the rest of the script.

The `;` between commands does the same thing.

> I'd actually prefer we write this script in Python. We can do it easily with the stdlib and we have access to python so expecting maintainers to master shell scripting in addition to python seems superfluous.

Makes sense.

**Comment by rgaudin** [comment]
> The `;` between commands does the same thing.

it does the opposite

**Comment by rgaudin** [comment]
> That's cool, how to use it? Do I get it right that it means you expect the cleanup script to live in this k8s repo? Any recommendation in term of location? (same folder than k8s cronjob?).

Yes, code in job description is hard to read and error-prone ; same goes for in-config maps. With this method, you get a clear, python file on the repo ; with your syntax coloring and linting if you wish. I find this easier to maintain.

Here's the repo for the image: https://github.com/rgaudin/docker-images/tree/main/debian
You are invited to port that to kiwix/container-images 🙃. I'm not sure why I kept it separate. Was probably a temp thing and I forgot.

**Comment by benoit74** [comment]
NOTA: for now the `INSTALL_SCRIPTS` is set to use a direct URL referencing a branch since this is not yet merged ... the chicken and egg problem ; I intend to migrate this to a github URL once this is merged and before deploying to cluster

File: zim/tmp-kiwix/cleanup-old-files.py
Comment (rgaudin) [reply zim/tmp-kiwix/cleanup-old-files.py:37]: And use the `Path().stat().st_mtime` to compare it directly. I don't think we need two loops

File: zim/tmp-kiwix/cleanup-old-files.py
Comment (rgaudin) [inline zim/tmp-kiwix/cleanup-old-files.py:66]: I'd probably loop on files_to_delete's parent recursively to build a list of paths to delete

File: zim/tmp-kiwix/cleanup-old-files.py
Comment (rgaudin) [inline zim/tmp-kiwix/cleanup-old-files.py:100]: You may want to catch your `Path().rmdir()` in case it's not empty for some reason (might be that someone else pushed something in) and not just crash the script

**Review by rgaudin — CHANGES_REQUESTED** [review]
Doesn't feel like @benoit74 😂

**Comment by benoit74** [comment]
> Doesn't feel like @benoit74 😂

But is it better or worse? (yes, I was not "alone" to write this ^^)

File: zim/tmp-kiwix/cleanup-old-files.py
Comment (rgaudin) [inline zim/tmp-kiwix/cleanup-old-files.py:47]: It's probably OK but this is just too complicated for me to understand… which I think defeats the maintenance goal.

Can you rewrite it in a less-elegant but simpler, easier to read way?
Also, I think we should bound this to the passed folder (which can end up being empty). We don't want to delete it.

Here's a version I find easier to understand

```py
def list_directories_to_delete(
    folder_path: Path, files_to_delete: list[Path]
) -> list[Path]:
    """List empty subdirectories or subdirectories containing only files to be deleted."""
    folder_candidates: set[Path] = set()

    # loop over files (deepest first) to add folders (no i/o needed)
    for file in sorted(files_to_delete, key=lambda p: len(p.parts), reverse=True):
        for parent in file.parents:
            # make sure we only add child folders
            if parent.is_relative_to(folder_path) and parent != folder_path:
                folder_candidates.add(parent)

    # loop over unique list of folder (i/o to get list of files for each folder)
    for folder in list(folder_candidates):
        # check if there are file on disk in folder that are not to be deleted
        if [1 for file in folder.iterdir() if file not in files_to_delete]:
            folder_candidates.remove(folder)

    return list(folder_candidates)
```

**Review by rgaudin — CHANGES_REQUESTED** [review]
Some more

File: zim/tmp-kiwix/cleanup-old-files.py
Comment (rgaudin) [inline zim/tmp-kiwix/cleanup-old-files.py:71]:
```suggestion
        print("\n".join(sorted(str(path) for path in files_to_delete)))
```

File: zim/tmp-kiwix/cleanup-old-files.py
Comment (rgaudin) [inline zim/tmp-kiwix/cleanup-old-files.py:73]:
```suggestion
        print("\n".join(sorted(str(path) for path in empty_directories_to_delete)))
```

File: zim/tmp-kiwix/cleanup-old-files.py
Comment (rgaudin) [reply zim/tmp-kiwix/cleanup-old-files.py]: Too lazy to check 😉 Anyway you code is clear now

**Review by rgaudin — APPROVED** [review]
Great! Feel free to use the implicit list comprehension, then rebase and merge.

**Pull request merged**
